pkg/gadgets/internal/socketenricher/bpf/sockets-map.h

```diff
@@ -0,0 +1,85 @@
+/* SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) OR Apache-2.0 */
+
+#ifndef SOCKETS_MAP_H
+#define SOCKETS_MAP_H
+
+#ifndef SOCKETS_MAP_IMPLEMENTATION
+
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+
+#endif
+
+#ifndef PACKET_HOST
+#define PACKET_HOST		0
+#endif
+
+#ifndef PACKET_OUTGOING
+#define PACKET_OUTGOING		4
+#endif
+
+#ifndef ETH_HLEN
+#define ETH_HLEN	14
+#endif
+
+const volatile __u32 current_netns = 0;
+
+unsigned long long load_byte(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.byte");
+unsigned long long load_half(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.half");
+unsigned long long load_word(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.word");
+
+#define L4_OFF (ETH_HLEN + sizeof(struct iphdr))
+
+typedef __u32 ipv4_addr;
+
+struct sockets_key {
+	__u32 netns;
+
+	// proto is IPPROTO_TCP(6) or IPPROTO_UDP(17)
+	__u16 proto;
+	__u16 port;
+};
```

Comment:
Thanks for the research!

We could probably add the family and the ip in the key to avoid clashes.

We have to be careful with IPv6 Addresses with Embedded IPv4 Addresses ([rfc4291 section-2.5.5](https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5)). Even if we don't care about IPv6 and the process uses IPv4, the kernel might internally convert it to IPv6. So we have to handle IPv6 addresses anyway. We can use the idea from [is_ipv4_mapped_ipv6()](https://github.com/weaveworks/tcptracer-bpf/blob/d329ae7fb6fa90e4389075347f0754de2127336f/tcptracer-bpf.c#L132-L133) to check that.

About SO_REUSEPORT, I suggest to overwrite the map entry with the last info. For ingress, socket filter programs are too early in the network stack to know which socket the kernel is going to choose. I don't think it is a big deal if we give the wrong socket data. The two sockets are likely to be in the same process. But if one socket is released earlier than the other one, we would remove the entry from the map too early and then we would not get any process data for that connection. Do you think it is a problem in practice?

Reply:
I think it's fine to assume that all sockets sharing a port belong to the same application, that's what SO_REUSEPORT was designed for.